main.go

```diff
@@ -3,117 +3,169 @@ package main
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"log"
+	"net"
+	"net/http"
 	"os"
 	"os/signal"
 	"path"
 	"syscall"
 
 	"github.com/hashicorp/vault-lambda-extension/config"
 	"github.com/hashicorp/vault-lambda-extension/extension"
+	"github.com/hashicorp/vault-lambda-extension/server"
 	"github.com/hashicorp/vault-lambda-extension/vault"
+	"github.com/hashicorp/vault/api"
 )
 
 const (
 	extensionName = "vault-lambda-extension"
 )
 
-var (
-	extensionClient = extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
-)
-
 func main() {
 	logger := log.New(os.Stderr, fmt.Sprintf("[%s] ", extensionName), log.Ldate|log.Ltime|log.LUTC)
 	ctx, cancel := context.WithCancel(context.Background())
+	extensionClient := extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
+	_, err := extensionClient.Register(ctx, extensionName)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	threadFinishedCh := make(chan struct{}, 2)
+	srv, err := initialiseExtension(logger, threadFinishedCh)
+	if err != nil {
+		logger.Fatal(err)
+	}
 
-	sigs := make(chan os.Signal, 1)
-	signal.Notify(sigs, syscall.SIGTERM, syscall.SIGINT)
+	shutdownChannel := make(chan struct{})
 	go func() {
-		s := <-sigs
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, syscall.SIGTERM, syscall.SIGINT)
+		select {
+		case s := <-interruptChannel:
+			logger.Printf("Received %s, exiting\n", s)
+		case <-shutdownChannel:
+			logger.Println("Received shutdown event, exiting")
+		}
+
 		cancel()
-		logger.Println("Received", s)
-		logger.Println("Exiting")
+		if err := srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners, or context timeout:
+			logger.Printf("HTTP server shutdown error: %s\n", err)
+		}
+		threadFinishedCh <- struct{}{}
 	}()
 
-	_, err := extensionClient.Register(ctx, extensionName)
-	if err != nil {
-		logger.Fatal(err)
-	}
+	processEvents(ctx, logger, extensionClient)
 
-	initialiseExtension(logger)
+	// Once processEvents returns, signal that it's time to shutdown.
+	shutdownChannel <- struct{}{}
 
-	processEvents(ctx, logger)
+	// Ensure we wait for the HTTP server to gracefully shut down.
+	// One signal from the HTTP server thread and one from the shutdown thread.
+	<-threadFinishedCh
+	<-threadFinishedCh
+	logger.Println("Graceful shutdown complete")
 }
 
-func initialiseExtension(logger *log.Logger) {
+func initialiseExtension(logger *log.Logger, threadFinishedCh chan struct{}) (*http.Server, error) {
```

Comment:
It may be appropriate to rename this to `RunExtension` or something, as the purpose changes from "writing the secrets to disk and quitting" to "run the proxy until things error or we shut down"